RawSpeed/LJpegPlain.cpp

```diff
@@ -557,376 +178,76 @@ void LJpegPlain::decodeScanLeft4_2_2() {
-    _ASSERTE((offset[slice]&0x0fffffff) < mRaw->pitch*mRaw->dim.y);
-    t_y++;
-    if (t_y == (frame.h - skipY)) {
-      t_y = 0;
-      t_x += slicesW[t_s++];
-    }
-  }
-  // We check the final position. If bad slice sizes are given we risk writing outside the image
-  if ((offset[slices-1]&0x0fffffff) >= mRaw->pitch*mRaw->dim.y) {
-    ThrowRDE("LJpegPlain::decodeScanLeft: Last slice out of bounds");
-  }
-
-  offset[slices] = offset[slices-1];        // Extra offset to avoid branch in loop.
-
-  slice_width = new int[slices];
-
-  // This is divided by comps, since comps pixels are processed at the time
-  for (uint32 i = 0 ; i <  slicesW.size(); i++)
-    slice_width[i] = slicesW[i] / COMPS;
-
-  if (skipX)
-    slice_width[slicesW.size()-1] -= skipX;
 
-  // First pixels are obviously not predicted
-  int p1;
-  int p2;
-  int p3;
-  ushort16 *dest = (ushort16*) & draw[offset[0] & 0x0fffffff];
-  ushort16 *predict = dest;
-  *dest++ = p1 = (1 << (frame.prec - Pt - 1)) + HuffDecode(dctbl1);
-  *dest++ = p2 = (1 << (frame.prec - Pt - 1)) + HuffDecode(dctbl2);
-  *dest++ = p3 = (1 << (frame.prec - Pt - 1)) + HuffDecode(dctbl3);
-
-  slice = 1;
-  uint32 pixInSlice = slice_width[0] - 1;
-
-  uint32 cw = (frame.w - skipX);
-  uint32 x = 1;                            // Skip first pixels on first line.
-  uint32 ch = (frame.h - skipY);
+      if (X_S_F == 1) { // will be optimized out
+        unroll_loop<N_COMP>([&](int i) {
+          *dest++ = p[i] += ht[i]->decodeNext(bitStream);
+        });
+      } else {
+        unroll_loop<Y_S_F>([&](int i) {
+          dest[0 + i*pixel_pitch] = p[0] += ht[0]->decodeNext(bitStream);
+          dest[3 + i*pixel_pitch] = p[0] += ht[0]->decodeNext(bitStream);
+        });
 
-  // Fix for Canon 80D mraw format.
-  // In that format, `frame` is 4032x3402, while `mRaw` is 4536x3024.
-  // Consequently, the slices in `frame` wrap around (this is taken care of by
-  // `offset`) and must be decoded fully (without skipY) to fill the image
-  if (mWrappedCr2Slices)
-    ch = frame.h;
+        dest[1] = p[1] += ht[1]->decodeNext(bitStream);
+        dest[2] = p[2] += ht[2]->decodeNext(bitStream);
 
-  for (uint32 y = 0;y < ch;y++) {
-    for (; x < cw ; x++) {
-      p1 += HuffDecode(dctbl1);
-      *dest++ = (ushort16)p1;
-
-      p2 += HuffDecode(dctbl2);
-      *dest++ = (ushort16)p2;
-
-      p3 += HuffDecode(dctbl3);
-      *dest++ = (ushort16)p3;
-
-      if (0 == --pixInSlice) { // Next slice
-        if (slice > slices)
-          ThrowRDE("LJpegPlain::decodeScanLeft: Ran out of slices");
-        uint32 o = offset[slice++];
-        dest = (ushort16*) & draw[o&0x0fffffff];  // Adjust destination for next pixel
-        if((o&0x0fffffff) > mRaw->pitch*mRaw->dim.y)
-          ThrowRDE("LJpegPlain::decodeScanLeft: Offset out of bounds");
-        _ASSERTE((o >> 28) < slicesW.size());
-        pixInSlice = slice_width[o>>28];
+        dest += 3 * sizeof(ushort16);
```

Comment:
```c++
*** CID 1397763:    (SIZEOF_MISMATCH)
/RawSpeed/LJpegPlain.cpp: 232 in RawSpeed::LJpegPlain::decodeN_X_Y<(int)2, (int)1, (int)1>()()
226               dest[3 + i*pixel_pitch] = p[0] += ht[0]->decodeNext(bitStream);
227             });
228
229             dest[1] = p[1] += ht[1]->decodeNext(bitStream);
230             dest[2] = p[2] += ht[2]->decodeNext(bitStream);
231
>>>     CID 1397763:    (SIZEOF_MISMATCH)
>>>     Adding "6UL /* 3 * sizeof (RawSpeed::ushort16) */" to pointer "dest" of type "RawSpeed::ushort16 *" is suspicious because adding an integral value to this pointer automatically scales that value by the size, 2 bytes, of the pointed-to type, "RawSpeed::ushort16".  Most likely, the multiplication by "sizeof (RawSpeed::ushort16)" in this expression is extraneous and should be eliminated.
232             dest += 3 * sizeof(ushort16);
233           }
234
235           pixInSlicedLine -= X_S_F;
236         }
237
/RawSpeed/LJpegPlain.cpp: 232 in RawSpeed::LJpegPlain::decodeN_X_Y<(int)3, (int)2, (int)2>()()
226               dest[3 + i*pixel_pitch] = p[0] += ht[0]->decodeNext(bitStream);
227             });
228
229             dest[1] = p[1] += ht[1]->decodeNext(bitStream);
230             dest[2] = p[2] += ht[2]->decodeNext(bitStream);
231
>>>     CID 1397763:    (SIZEOF_MISMATCH)
>>>     Adding "6UL /* 3 * sizeof (RawSpeed::ushort16) */" to pointer "dest" of type "RawSpeed::ushort16 *" is suspicious because adding an integral value to this pointer automatically scales that value by the size, 2 bytes, of the pointed-to type, "RawSpeed::ushort16".  Most likely, the multiplication by "sizeof (RawSpeed::ushort16)" in this expression is extraneous and should be eliminated.
232             dest += 3 * sizeof(ushort16);
233           }
234
235           pixInSlicedLine -= X_S_F;
236         }
237
/RawSpeed/LJpegPlain.cpp: 232 in RawSpeed::LJpegPlain::decodeN_X_Y<(int)3, (int)2, (int)1>()()
226               dest[3 + i*pixel_pitch] = p[0] += ht[0]->decodeNext(bitStream);
227             });
228
229             dest[1] = p[1] += ht[1]->decodeNext(bitStream);
230             dest[2] = p[2] += ht[2]->decodeNext(bitStream);
231
>>>     CID 1397763:    (SIZEOF_MISMATCH)
>>>     Adding "6UL /* 3 * sizeof (RawSpeed::ushort16) */" to pointer "dest" of type "RawSpeed::ushort16 *" is suspicious because adding an integral value to this pointer automatically scales that value by the size, 2 bytes, of the pointed-to type, "RawSpeed::ushort16".  Most likely, the multiplication by "sizeof (RawSpeed::ushort16)" in this expression is extraneous and should be eliminated.
232             dest += 3 * sizeof(ushort16);
233           }
234
235           pixInSlicedLine -= X_S_F;
236         }
237
/RawSpeed/LJpegPlain.cpp: 232 in RawSpeed::LJpegPlain::decodeN_X_Y<(int)4, (int)1, (int)1>()()
226               dest[3 + i*pixel_pitch] = p[0] += ht[0]->decodeNext(bitStream);
227             });
228
229             dest[1] = p[1] += ht[1]->decodeNext(bitStream);
230             dest[2] = p[2] += ht[2]->decodeNext(bitStream);
231
>>>     CID 1397763:    (SIZEOF_MISMATCH)
>>>     Adding "6UL /* 3 * sizeof (RawSpeed::ushort16) */" to pointer "dest" of type "RawSpeed::ushort16 *" is suspicious because adding an integral value to this pointer automatically scales that value by the size, 2 bytes, of the pointed-to type, "RawSpeed::ushort16".  Most likely, the multiplication by "sizeof (RawSpeed::ushort16)" in this expression is extraneous and should be eliminated.
232             dest += 3 * sizeof(ushort16);
233           }
234
235           pixInSlicedLine -= X_S_F;
236         }
237

```